### ISSUE TITLE
Reduce hang in visiting next email

### ISSUE THREAD

**Issue by bradyt** [djcb/mu#1217]
## Expected or desired behavior

Hold down the `n` key and flip quickly through emails.

This youtube video suggests that mutt can handle this [Using Mutt deleting spam in realtime - YouTube](https://www.youtube.com/watch?v=Vq6U38WHpsw). I've reproduced in mutt, with html rendering via w3m, and autoview for all html emails, and there is still no real noticeable hang.

## Actual behavior

For just a few key presses, there is a slightly noticeable amount of "Waiting for message..." before messages appear rendered.

When holding down the `n` key, there is also intermittent hanging before cursor in headers pane moves.

## Steps to reproduce

Occurs even for

```
(setq mu4e-view-prefer-html nil
      mu4e-html2text-command (lambda (msg) ""))
```

## Versions of mu, mu4e/emacs, operating system etc.

```
mu (mail indexer/searcher) version 1.0
GNU Emacs 25.3.1 (x86_64-apple-darwin17.4.0, NS appkit-1561.20 Version 10.13.3 (Build 17D102)) of 2018-03-30
```

**Comment by djcb** [comment]
I don't notice any "hang"; how many millisecs are we talking?

**Comment by bradyt** [comment]
I now think this is two issues, not sure if we should split this. Again, I have key repeat rate up, so not sure if you are taking that step.

# Issue 1

TL;DR: Part of my issue is found to be `evil-mode`'s `normal` state in `fundamental-mode`, for `Waiting for message...`.

When holding down the `n` key, part of the issue is associated with a message, from `*Messages*` buffer, `user-error: "mark": pattern not found`, then `M-x toggle-debug-on-error` doesn't quite catch that one until a suggestion on #emacs to try `(setq debug-ignored-errors (delq 'user-error debug-ignored-errors))`. Then I note stack trace has `evil-search-next`, ponder how that's being called, and notice the `Waiting for message...` buffer is in `fundamental-mode` with `evil-mode`'s `normal` state.

Maybe the `Waiting for message...` buffer should already be in `mu4e:view` mode? This is what you see when starting `gnus`, but then, that start up page is not something we are flipping through. If I look at how `gnus` traverses articles, it doesn't move cursor in headers buffer until view has updated. Maybe that sort of hang is not desired. I'm not sure what `mutt` does, I see flashing in their version of echo area.

# Issue 2

The `Waiting for message...`, and time it takes in that status, you asked how many milliseconds, I'm not sure how to measure that. Also, I suspect the experience may be slightly different on `macOS` vs `GNU/Linux`. I would have to set that up at some point. Maybe I can setup a mailing list read from mu4e, so I can do an asciinema for example.

**Comment by bradyt** [comment]
Looking at the function `mu4e~headers-get-loading-buf` defined at [mu4e-headers.el#L1561-L1574](https://github.com/djcb/mu/blob/9ecc610e141bde6fcf0961715fabfe71f2ad80aa/mu4e/mu4e-headers.el#L1561-L1574), one #emacs'r suggests this should be in `special-mode`, or derived from it, not from `fundamental-mode`. Perhaps if there's a special mu4e waiting mode, it would be easier to rebind `n` and `p`. Right now, my horrible horrible fix are the two lines with `evil-local-set-key`, in

```elisp
(defun mu4e~headers-get-loading-buf ()
  "Get a buffer to give feedback while loading a message view."
  (unless (buffer-live-p mu4e~headers-loading-buf)
    (setq mu4e~headers-loading-buf
      (get-buffer-create " *mu4e-loading*")))
  (with-current-buffer mu4e~headers-loading-buf
      (let ((inhibit-read-only t))
	(erase-buffer)
	(local-set-key (kbd "q") (if (eq mu4e-split-view 'single-window)
				     'kill-buffer
				   'kill-buffer-and-window))
	(evil-local-set-key 'normal (kbd "n") 'ignore)
	(evil-local-set-key 'normal (kbd "p") 'ignore)
	(insert (propertize "Waiting for message..."
		  'face 'mu4e-system-face 'intangible t))))
  mu4e~headers-loading-buf)
```

**Comment by djcb** [comment]
Hmm, it seems `evil`-specific then. At least for debugging, I'd recommend trying to disable `evil` (and any 3rd party packages as much as possible), and see if you can still reproduce.

**Comment by bradyt** [comment]
Sorry, I meant to say, "Issue 2" remains after disabling evil-mode. My claim is there are now two issues.

**Comment by djcb** [comment]
The first issue, I guess you can best take up with the `evil` maintainers. 

For the second, I can't reproduce that; even without precise measurements, perhaps you can make an estimate at least... does it take more than 100ms? 500ms? a second? Here, I can typically not even _see_ the `Waiting for message...` since it's so fast.

**Comment by bradyt** [comment]
Did you try increasing the key repeat rate?

**Comment by bradyt** [comment]
For issue 1, an Evil maintainer @wasamasa suggested I use `evil-buffer-regexps`, so a better fix is now

``` elisp
(add-to-list 'evil-buffer-regexps '("^ \\*mu4e-loading\\*"))
```

Let's consider issue 1 closed! :)

**Comment by bradyt** [comment]
For issue 2, as I flip through emails, even one at a time, I can get delays averaging maybe a fraction of a second, and some less frequent at half a second. But this is getting quite subjective. If you like, we can close this issue, until we get some video. I've made an extra email account, I'll try to make time to subscribe to some lists, and perhaps make an `asciinema` video. I should also check in `GNU/Linux` at some point to see if _I_ find a significant difference that way.

**Comment by thblt** [comment]
FWIW I've tried running the memory profiler while flipping through a few emails (exactly 100) at full speed.  Memory allocation goes over 2GB, half of which in `mu4e-hide-other-mu4e-buffers`, the other half in visual-fill-column (and can probably be ignored then since it's unrelated to mu4e).  This is with (setq mu4e-html2text-command (lambda (&rest _) "")) to avoid HTML conversion noise.

```
- command-execute                                       2,326,725,258  62%
 - call-interactively                                   2,326,707,590  62%
  - funcall-interactively                               2,326,707,590  62%
   - mu4e-view-headers-prev                             2,294,510,834  61%
    - mu4e~headers-move                                 2,294,467,746  61%
     - mu4e-headers-view-message                        2,294,384,038  61%
      - mu4e~headers-redraw-get-view-window             2,284,592,406  61%
       - mu4e-hide-other-mu4e-buffers                   1,264,475,133  34%
        - delete-window                                 1,264,469,853  34%
         - run-window-configuration-change-hook             7,555,605   0%
          + thblt/update-window-position-parameters         7,162,025   0%
          + which-key--hide-popup-on-frame-size-change        369,292   0%
          + window--adjust-process-windows                     24,288   0%
       - eval                                           1,020,106,713  27%
        - split-window-horizontally                     1,020,106,713  27%
         - split-window                                 1,020,106,713  27%
          + run-window-configuration-change-hook            9,524,337   0%
       + switch-to-buffer                                       4,224   0%
      + switch-to-buffer                                    9,727,668   0%
      + mu4e~headers-get-loading-buf                           35,728   0%
      + mu4e~proc-view                                         11,348   0%
      + select-window                                           4,224   0%
     + mu4e~headers-highlight                                  57,308   0%
     + walk-windows                                            22,176   0%
      mu4e~headers-goto-docid                                  25,136   0%
    + select-window                                             2,112   0%
   + counsel-M-x                                           17,854,366   0%
   + mu4e~view-quit-buffer                                  7,552,105   0%
   + mu4e-headers-view-message                              6,555,506   0%
   + next-line                                                 87,464   0%
   + previous-line                                             62,304   0%
   + handle-focus-out                                          50,494   0%
   + profiler-report-toggle-entry                              20,789   0%
   + handle-focus-in                                            1,056   0%
- mu4e~proc-filter                                      1,369,109,267  36%
 - mu4e~headers-view-handler                            1,361,017,096  36%
  - mu4e-view                                           1,359,014,181  36%
   - mu4e-view-mode                                     1,008,576,276  27%
    - run-mode-hooks                                    1,008,516,084  27%
     - apply                                            1,008,111,904  27%
      - run-hooks                                       1,008,111,904  27%
       - #<lambda 0xa3bb0f50d>                          1,007,075,922  27%
        - visual-fill-column-mode                       1,007,034,738  27%
         - visual-fill-column-mode--enable              1,007,012,562  27%
          - visual-fill-column--adjust-window           1,007,007,282  27%
           - visual-fill-column--set-margins            1,006,844,073  27%
              visual-fill-column--window-max-text-width         3,168   0%
            add-hook                                            3,168   0%
         + called-interactively-p                              12,672   0%
           run-hooks                                            3,168   0%
        + visual-line-mode                                     30,624   0%
       + mu4e-alert-update-mail-count-modeline              1,008,526   0%
       + beginend-message-mode                                 19,008   0%
         #<compiled 0xde8ae9>                                   4,224   0%
     + run-hooks                                              404,180   0%
    + special-mode                                             34,848   0%
   + mu4e~view-make-urls-clickable                        185,667,476   5%
   + mu4e~view-show-images-maybe                          130,115,676   3%
   + mu4e-view-message-text                                21,277,164   0%
   + switch-to-buffer                                       9,332,809   0%
   + mu4e~fontify-cited                                       935,694   0%
   + undo-auto--undoable-change                                 2,104   0%
   string-match                                                 4,532   0%
   decode-coding-string                                         2,182   0%
   mu4e-info-handler                                            1,056   0%
- redisplay_internal (C function)                           8,066,559   0%
 - eval                                                     6,054,320   0%
  + when                                                    3,084,233   0%
  + moody-tab                                               2,448,424   0%
  + thblt/mode-line-sep                                       328,572   0%
 + #<compiled 0x438425>                                        45,408   0%
   auto-compose-chars                                           6,341   0%
+ sp--save-pre-command-state                                   52,800   0%
+ selected-global-mode-check-buffers                           22,176   0%
+ #<lambda 0x3f422448c06>                                      21,120   0%
+ timer-event-handler                                          19,410   0%
+ global-font-lock-mode-check-buffers                          16,896   0%
+ yas-global-mode-check-buffers                                15,840   0%
+ smartparens-global-mode-check-buffers                        15,840   0%
+ global-undo-tree-mode-check-buffers                          10,560   0%
+ show-smartparens-global-mode-check-buffers                    9,504   0%
+ magit-auto-revert-mode-check-buffers                          5,280   0%
+ mu4e~update-sentinel-func                                     2,692   0%
+ eldoc-schedule-timer                                          2,112   0%
+ mode-local-post-major-mode-change                             2,112   0%
+ yas--post-command-handler                                     1,216   0%
  ...                                                               0   0%
```
```

**Comment by djcb** [comment]
Oh, that's interesting... not sure why it would allocate so much memory. How did you do the profiling?

**Comment by thblt** [comment]
> How did you do the profiling?

I'm not sure I understand what exactly it is you're asking, but I did
follow the following steps:

 - Show exactly 100 headers in mu4e headers view
 - Open the first mail
 - Run the RAM profiler using `M-x profiler-start`
 - Scroll through all mails by holding the `n` key
 - When at the last mail, run `M-x profiler-report`

Please tell me if you need any more information

**Comment by djcb** [comment]
There's a new (experimental) message-view... when you set `(setq mu4e-view-use-gnus t)` before starting mu4e. Does that help for the speed?

**Comment by bradyt** [comment]
No, no noticeable change. I thought of something though. I noticed in `magit`, at first I thought it was strange, that you could traverse the log faster than the preview updates. But now it seems like a great idea, to be able to hold down `n` to get to a later commit faster, and preview gracefully updates when it can.

However, the cursor stays in the log buffer. I believe Gnus does this too (though it prevents cursor from moving until preview updates). Would this idea be feasible here, to allow to move down list of emails quickly, and let preview catch up when it can?

**Comment by djcb** [comment]
Maybe... but since I can't even reproduce the problem on older machines (and there seems to be no problem with `emacs -nw`), it seems the problem might be somewhere else. I.e. I would need to see where exactly the time is spent (an GUI vs `-nw`).

**Comment by bradyt** [comment]
Maybe we see an improvement in UI, if ` *mu4e-loading*` buffer uses a mode derived from `special-mode`, with it's own keymap, then `n` and `p` can be bound properly there, rather than `self-insert-command` and seeing

```
funcall-interactively: Buffer is read-only: #<buffer  *mu4e-loading*> [17 times]
```

In echo area and messages buffer.

**Comment by djcb** [comment]
We now use a special-mode derived loading buffer. We can't speed things up much more given the current setup, but there are some ideas for when we move to gnus-based view to avoid mu-server round-trips.
